### PR TITLE
duration param

### PR DESCRIPTION
**Pull request by BanzaiTokyo**

### CODE REVIEW

File: packages/hardhat/contracts/LotteryDeployer.sol [hunk 1]
```diff
@@ -17,9 +17,19 @@ contract LotteryDeployer is Ownable {
 
   /// @notice Creates a new Lottery contract
   /// @dev Deploys a new Lottery contract and stores the address in the lotteries mapping
+  function createLottery(uint256 duration) external {
+
+    require(duration <= 14 days);
```
Comment (gurobokum) [inline packages/hardhat/contracts/LotteryDeployer.sol:22]: Could you please provide error message

File: packages/hardhat/contracts/LotteryDeployer.sol [hunk 2]
```diff
@@ -17,9 +17,19 @@ contract LotteryDeployer is Ownable {
...
+
+    Lottery lottery = new Lottery(block.timestamp + duration);
```
Comment (gurobokum) [inline packages/hardhat/contracts/LotteryDeployer.sol:24]: I'm not sure but looks like by design would be better to calculate timestamp in Lottery itself

File: packages/hardhat/contracts/LotteryDeployer.sol [hunk 3]
```diff
@@ -17,9 +17,19 @@ contract LotteryDeployer is Ownable {
...
+    uint256 index = lotteryCount[msg.sender];
+    lotteries[msg.sender][index] = address(lottery);
+    lotteryCount[msg.sender]++;
+  }
+
   function createLottery() external {
```
Comment (gurobokum) [inline packages/hardhat/contracts/LotteryDeployer.sol:30]: Needs a parama

File: packages/hardhat/contracts/Lottery.sol [hunk 4]
```diff
@@ -42,6 +42,9 @@ contract Lottery is AccessControl {
 
         //TODO: make sure if the msg.sender is indeed the starter and not the proxy contract
         _grantRole(STARTER_ROLE, msg.sender);
+
+        //TODO: replace Vitalik's wallet with a more appropriate address.
+        _grantRole(DEVELOPER_ROLE, address(0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045));
```
Comment (gurobokum) [inline packages/hardhat/contracts/Lottery.sol:47]: I guess we need to pass this information from LotteryDeployer

**Review by gurobokum — CHANGES_REQUESTED** [review]

File: packages/hardhat/contracts/Lottery.sol [hunk 5]
```diff
@@ -37,11 +37,18 @@ contract Lottery is AccessControl {
     /// @dev List of bet slots
     address[] _slots;
 
-    constructor(uint256 _closingTime)  {
-        betsClosingTime = _closingTime;
+    constructor(uint256 duration)  {
+        if (duration <= 0 || duration > 14 days) {
```
Comment (gurobokum) [inline packages/hardhat/contracts/Lottery.sol:41]: I think uint couldn't be negative

Comment (gurobokum) [reply packages/hardhat/contracts/Lottery.sol:41]: Also I prefer to send error message about duration longer than 14 days, it's more explicit

**Review by gurobokum — CHANGES_REQUESTED** [review]

File: packages/hardhat/contracts/Lottery.sol (hunk 5 above)
Comment (gurobokum) [reply packages/hardhat/contracts/Lottery.sol:41]: Could you please also fix lint error

File: packages/nextjs/app/frames/lottery/route.tsx [hunk 6]
```diff
@@ -12,14 +13,15 @@ const handler = frames(async ({ searchParams: { chainId, address }, url: { href
         style={{ background: "linear-gradient(to right, rgb(45, 212, 191), rgb(254, 240, 138))" }}
       >
         <div>Participate in Lottery</div>
-        <img
+        <Image
```
Comment (gurobokum) [inline packages/nextjs/app/frames/lottery/route.tsx:16]: It should be img for frame, also it's warning not error

**Review by gurobokum — CHANGES_REQUESTED** [review]
Linter still fails

File: packages/nextjs/app/frames/lottery/route.tsx (hunk 6 above)
Comment (BanzaiTokyo) [reply packages/nextjs/app/frames/lottery/route.tsx]: reverted. But it kinda looks like an error
Screenshot 2024-05-02 at 12 09 01

File: packages/hardhat/contracts/Lottery.sol [hunk 7]
```diff
@@ -35,14 +35,32 @@ contract Lottery is AccessControl {
...
+        if (duration == 0) {
+            betsClosingTime = block.timestamp + 14 days;
+        } else {
+            betsClosingTime = block.timestamp + duration;
+        }
+
+        if (_starterFee == 0) {
+            starterFee = 500000000000000;
+        } else {
+            starterFee = _starterFee;
```
Comment (gurobokum) [inline packages/hardhat/contracts/Lottery.sol:54]:
```suggestion
       starterFee = _starterFee;
```

**Review by gurobokum — APPROVED** [review]

**Pull request merged**